### ISSUE TITLE
Elements not clickable after mouse hotplug / module reload

### ISSUE THREAD

**Issue by merlinschumacher** [awesomeWM/awesome#415]
When i reload the i2c_hid module or plug a usb mouse in or out awesomewm won't react to any mouse input anymore. GTK-programs still work fine. Even those in sitting the tray. This happens with my customized rc.lua and with the default one. 
I'm using 3.5.6 on Arch.

**Comment by merlinschumacher** [comment]
A restart of awesome inside the xsession also wont help. Only a logout. Awesome is still reacts fine to keyboard commands.

**Comment by pimotte** [comment]
I have encountered this issue as well, and it seems to be related to qt4 applications. If I replug my mouse while quasselclient or owncloud-client is open, awesome will no longer respond to clicks.
At most one of the applications will swallow the mouse events.

Scenarios that may help:

Scenario 1:
1. First start quasselclient, then owncloud
2. Unplug mouse, plug mouse in. [awesome does not respond to clicks]
3. Close quasselclient. [awesome does respond to clicks]
4. Replug mouse. [awesome does not respond to clicks]
5. Close owncloud-client. [awesome does respond to clicks]

Scenario 2:
1. First start quasselclient, then owncloud-client
2. Unplug mouse, plug mouse in. [awesome does not respond to clicks]
3. Close owncloud. [awesome does respond *not* to clicks]
4. Close quasselclient. [awesome does respond to clicks]

Scenario 3:
1. First start owncloud-client, then quasselclient
2. Unplug mouse, plug mouse in. [awesome does not respond to clicks]
3. Close owncloud. [awesome does respond *not* to clicks]
4. Close quasselclient. [awesome does respond to clicks]

Scenario 4:
1. Start owncloud-client
2. Replug mouse [awesome does not respond to clicks]
3. Close owncloud [awesome does respond to clicks]

**Comment by merlinschumacher** [comment]
@pimotte i tried it and it fixed my issue. closing and reopening owncloud worked like a charm

**Comment by Elv13** [comment]
I just updated my Gentoo, it now also happen to me (v3.5.6 too, it worked before)

**Comment by psychon** [comment]
@Elv13: Any qt4 applications that you can blame?

**Comment by jinks** [comment]
I can confirm the same behaviour for MultiMC5 (a Qt5 Minecraft launcher) under awesome 3.5.6 from the Arch Linux package.
Closing MultiMC fixes it, restarting awesome (while leaving MultiMC open) does not. Resizing tiles, focus-follows-mouse and so on still work, it seems only the wibar is affected (and interacting with the "desktop background").

Observation on this bug: It seems that so far all instances of this bug that have been reported are about applications that include a systray widget. No idea if that's relevant.

**Comment by merlinschumacher** [comment]
The owncloud client provokes exactly this behavior also
Am 07.11.2015 7:43 nachm. schrieb "Jinks" <notifications@github.com>:

**Comment by Elv13** [comment]
Yes, as a KDE user/dev, I do have a lot of Qt apps running. I upgraded my IRC client to use Qt5 (the last Qt4 app I had open at that time) and the problem stopped. So for me, it is indeed Qt4 only. 

* The issue cannot be reproduced with other WMs
* Awesome is **not** deadlocked, the keybindings still work and inputs get to clients (all of them, including the Qt4 app)
* Restarting Awesome does not help

**Comment by dubhater** [comment]
I have this problem too.

The wibox no longer reacts to touchpad clicks after plugging or unplugging my USB headphones while a Qt5 application is running. Things go back to normal if I close the Qt5 application. Sometimes it takes a second or two after (un)plugging the headphones before the wibox stops responding to clicks.

Qt5 applications tested: Calibre, Qt Creator, Transmission.
Qt4 application tested without triggering the problem: VLC. (I don't have others.)

Awesome is 3.5.6, Qt5 is 5.5.0, Qt4 is 4.8.7, all from the official Arch Linux packages.

**Comment by merlinschumacher** [comment]
Do your headphones have buttons? I'm asking because this might be input
related

--
Merlin Schumacher
Sedanstraße 28
31134 Hildesheim

+491716714810

Zebrabutter - Das Beste im Netz seit 1679
http://www.zebrabutter.net

Der Popkulturkalender - Jeden Tag ein Häppchen Popkultur
http://popkulturkalender.blogspot.com

PGP Fingerprint: 8201 8716 EF71 081B 901C  F8AD 3738 FEE8 C06C 57DA

2015-11-12 18:53 GMT+01:00 dubhater <notifications@github.com>:

**Comment by N0T3P4D** [comment]
I can confirm the problem running Gentoo with awesome 3.5.6.
Qt 4.8.7 work flawlessly, as does Qt5 5.4.2. Upgrading to Qt 5.5.1 (the next available packaged version) and running a Qt5 app triggers the problem, downgrading solves it.
I can reproduce it by (un)plugging a mouse, keyboard or YubiKey.

**Comment by dubhater** [comment]
Yes, my headphones have buttons: volume up, volume down, mute. xev reacts to the volume buttons.

**Comment by psychon** [comment]
A summary of the results we have so far:

- Awesome version does not seem to matter (well, everyone uses 3.5.6 or does not mention their version)
- Qt versions *without* this problem: 4.8.7 (@N0T3P4D, @dubhater), 5.4.2 (@N0T3P4D)
- Qt versions *with* this problem: 5.5.0 (@dubhater), 5.5.1 (@N0T3P4D)
- People who can reproduce this problem but didn't mention their Qt version: @merlinschumacher, @pimotte, @Elv13, @jinks 
- People who so far didn't reproduce this problem @psychon (and others)

Looking at the changelog for Qt 5.5 (http://doc.qt.io/qt-5/whatsnew55.html), only one thing caught my eye:

> Qt Platform Abstraction (QPA)
> [...]
> Added a libinput plugin.

Does anyone know if this libinput plugin is actually used here? However, that would not explain why Qt 4 also triggers this problem. And why apparently only awesome "suffers".
Did someone look into xev output for mouse clicks? Do the ButtonPress and ButtonRelease events have a "weird" `state` field?

**Comment by merlinschumacher** [comment]
Am 16.11.2015 um 16:21 schrieb Uli Schlachter:
> A summary of the results we have so far:
>
>   * Awesome version does not seem to matter (well, everyone uses 3.5.6
>     or does not mention their version)
>   * Qt versions /without/ this problem: 4.8.7 (@N0T3P4D
>     <https://github.com/N0T3P4D>, @dubhater
>     <https://github.com/dubhater>), 5.4.2 (@N0T3P4D
>     <https://github.com/N0T3P4D>)
>   * Qt versions /with/ this problem: 5.5.0 (@dubhater
>     <https://github.com/dubhater>), 5.5.1 (@N0T3P4D
>     <https://github.com/N0T3P4D>)
>   * People who can reproduce this problem but didn't mention their Qt
>     version: @merlinschumacher <https://github.com/merlinschumacher>,
>     @pimotte <https://github.com/pimotte>, @Elv13
>     <https://github.com/Elv13>, @jinks <https://github.com/jinks>
>   * People who so far didn't reproduce this problem @psychon
>     <https://github.com/psychon> (and others)
>
> Looking at the changelog for Qt 5.5
> (http://doc.qt.io/qt-5/whatsnew55.html), only one thing caught my eye:
>
>     Qt Platform Abstraction (QPA)
>     [...]
>     Added a libinput plugin.
>
> Does anyone know if this libinput plugin is actually used here? However,
> that would not explain why Qt 4 also triggers this problem. And why
> apparently only awesome "suffers".
> Did someone look into xev output for mouse clicks? Do the ButtonPress
> and ButtonRelease events have a "weird" |state| field?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/awesomeWM/awesome/issues/415#issuecomment-157065056>.
>
i use awesome 3.5.6 and qt 5.5.1

-- 
Merlin Schumacher
Sedanstraße 28
31134 Hildesheim

+491716714810

Zebrabutter - Das Beste im Netz seit 1679
http://www.zebrabutter.net

Der Popkulturkalender - Jeden Tag ein Häppchen Popkultur
http://popkulturkalender.blogspot.com

PGP Fingerprint: 8201 8716 EF71 081B 901C  F8AD 3738 FEE8 C06C 57DA

**Comment by pimotte** [comment]
Also awesome 3.5.6 and qt 5.5.1 here.

**Comment by psychon** [comment]
Having just found https://github.com/qtproject/qtbase/commit/53d289ec4c0f512a3475da4bbf1f940cd6838ace: Could someone try if setting `QT_XCB_NO_XI2_MOUSE=1` in Qt's environment makes this problem go away?

**Comment by dubhater** [comment]
Indeed it does.

**Comment by psychon** [comment]
I tried reproducing this with Qt build from git (version `v5.6.0-alpha1-1083-gf7ab199`, via the `examples/gui/analogclock/` program) and, well, everything works. I cannot reproduce this problem. At least I figured out that setting `QT_LOGGING_RULES="qt.qpa.input*=true"` in the env gets us some debug output, but there is nothing interesting in there for me: https://gist.github.com/psychon/910518fc62dfe23d5aa2

**Comment by Elv13** [comment]
By the way, an important details I missed, Awesome does get inputs, but only for clients, not wiboxes. So, mouse focus and titlebars buttons works

**Comment by Elv13** [comment]
Looking at Qt platform support plugin, maybe we should just support `libinput`. It seem simple enough (as an API, I am not commenting on how to implement it in Awesome). The biggest issue is the BSD/OSX ports are in early stages, so XCB would have to be kept as an alternative backend. Plus, it has better gesture support than Awesome for almost free. So gesture based window managing (like OSX) could be implemented.

**Comment by blueyed** [comment]
I am seeing this now suddenly, too.
It is caused by https://github.com/hluk/CopyQ for me, which sits in the systray.
If I exit it, the clicks go through to awesome's tasklist again.
Restarting it will not trigger it again.

So it might be related to the time / order / way it gets started.

I am using copyq since a while, and the issue popped up only now.
One thing to mention also is that the instance I just quit had different DPI settings (96) than the one I am setting up manually (176).  That was not the case before.
Will have to investigate further, but just wanted to leave this here for now.

**Comment by psychon** [comment]
A quick idea: Figure out the window id of the wibox via `xwininfo -root -tree`, then run `xev -id 0xTHE-WINDOW-ID` to see what kind of window events awesome would get. Does it really get none at all? Is some weird modifier present?

**Comment by blueyed** [comment]
It just happened again.
Awesome seems to get literally none events at all.

Tried it with these ids, which should correspond to the wibox tasklist entry of
Firefox:

     0x2000148 (has no name): ()  1920x1047+0+33  +0+33
        1 child:
        0x3600295 "Some Firefox window title - Vimperator": ("Navigator" "Firefox")  1920x1047+0+0  +0+33
           1 child:
           0x3600296 (has no name): ()  1x1+-1+-1  +-1+32

**Comment by blueyed** [comment]
I would say upgrading CopyQ from 2.4.9 to 2.5.0 triggered the issue here, but I still cannot say how it gets triggered exactly.
For reference, the diff between copyq 2.4.9 and 2.5.0: https://github.com/hluk/CopyQ/compare/v2.4.9...v2.5.0

**Comment by blueyed** [comment]
This one looks suspicious: https://github.com/hluk/CopyQ/commit/cfc1f31f395554eb5ab643959297836b80590652

**Comment by Elv13** [comment]
Well, one way for "force" the problem with any app (including Qt) programatically

1) Create a QMenu and add a couple of actions
2) `m->show()` (or `exec()`)
3) Catch the `hovered` signal and do `sleep(1000)`
4) Move the mouse over the menu

This will reproduce the result of this bug, but I am far from convinced it is the cause. There should be no input grabber that magically trigger when resuming. And this would freeze any WMs, not just Awesome.

**Comment by psychon** [comment]
Oh, that might make sense. If you want me to guess: Qt grabs all mouse input while a menu is open. The same will happen if you run e.g. gedit from vt1, open the "File" menu, switch back to vt1 and press Ctrl-Z. Gedit will have grabbed mouse input and nothing else gets it.
(Also, normally a menu closes when you click somewhere else, no?)
However, I don't think that behaviour is related to this issue and I guess it will happen with any WM. Does the same also happen with `QT_XCB_NO_XI2_MOUSE=1`?

**Comment by psychon** [comment]
@blueyed Can you try if downgrading to 2.4.9 fixes the issue for you? Perhaps do a bisect?

**Comment by blueyed** [comment]
@psychon 
Will try later.  As for now, I've noticed that when this happens there appears to be some text artifacts in the background for the copyq systray icon.  Disabling the systray option with copyq does not fix it btw.

**Comment by gaudenz** [comment]
I have the same issue with owncloud and keepassx on QT 5.5.1 from Debian and awesome 3.5.6 also from Debian. I can confirm that starting with QT_XCB_NO_XI2_MOUSE=1 fixes the issue. Is there anything that can be done on the awesome side against this or is this a QT bug?

**Comment by gaudenz** [comment]
As an additional info the issue is not there immediately after the application starts but I can trigger it by unplugging my yubikey (which is among other things an input device). So it seems to be input device related. Applications other than awesome don't seem to be affected. Mouse events work just fine there even if awesome does not process any events anymore.

**Comment by psychon** [comment]
Could someone test if adding `| XCB_EVENT_MASK_BUTTON_PRESS | XCB_EVENT_MASK_BUTTON_RELEASE ` [here](https://github.com/awesomeWM/awesome/blob/742cbd712273f51a48e7d239567f5b31306e0295/objects/drawin.c#L349) helps?

**Comment by N0T3P4D** [comment]
@psychon 
I've upgraded to Qt 5.5.1 (without libinput support for qtgui) again and changed the source of awesome 3.5.6 according to your suggestion. With the exception of right-clicking on the desktop, this seems to fix the problem.
The workaround with QT_XCB_NO_XI2_MOUSE=1 works here, too.

**Comment by gaudenz** [comment]
Uli Schlachter <notifications@github.com> writes:

> Could someone test if adding `| XCB_EVENT_MASK_BUTTON_PRESS |
> XCB_EVENT_MASK_BUTTON_RELEASE `
> [here](https://github.com/awesomeWM/awesome/blob/742cbd712273f51a48e7d239567f5b31306e0295/objects/drawin.c#L349)
> helps?

Yes this fixes the problem for me. I did not change anything else (same
QT version) and after this change the QT_XCB_NO_XI2_MOUSE=1 workaround
is no longer needed.

Gaudenz

**Comment by blueyed** [comment]
@psychon 
Should the change get added to awesome, or is it considered to be a (bad) workaround?

**Comment by psychon** [comment]
@blueyed Did that in https://github.com/awesomeWM/awesome/commit/a86d1137eeaa70e6cc138e2a0b89f2e6196a08f7
@everyone How much does the above help? @N0T3P4D wrote that right-clicking on the root window is (still) broken with this patch? If so, then I guess that this just fixes one of the symptoms and Qt asking for button events somehow still has priority over us...

**Comment by psychon** [comment]
Ok, I think I figured this out. The above commit should "accidentally" fix it for all drawins, but I cannot do anything about mouse clicks on the root window (except for making awesome use the XInput extension, but "because Qt" isn't really a good reason for that).

Could someone report this to Qt's bug tracker and give the link here? I'd do that, but someone who can reproduce this might be more useful. :-)
I'm looking into how to fix this properly and will explain what is going on on the Qt bug.

**Comment by psychon** [comment]
Completely untested patch for Qt: https://gist.github.com/psychon/76cdb38d079ee976d35d
Another nice thing that someone could do is to rebuild Qt with the above patch and see if it works (does awesome get mouse clicks on the root window and in the wibox? Does Qt get mouse events with mice that have been "always there" and also with mice that were connected only after some Qt app started?)
The patch should make Qt only select for XInput2 events on mouse hotplug on the windows that it also asked for during "normal operation" (= windows that it created itself; not the root window).

**Comment by N0T3P4D** [comment]
@psychon I've tested the Qt patch with additional debug output and unpatched awesome. Unfortunately, the problem persists and I can see the additional lines with low numbers.

**Comment by psychon** [comment]
Thanks for testing. Seems like I didn't understand how Qt ends up including the root window in its event list...

**Comment by N0T3P4D** [comment]
I've filed https://bugreports.qt.io/browse/QTBUG-49952

**Comment by psychon** [comment]
Thanks, I left a large-ish comment and subscribed to updates there.

**Comment by Malvineous** [comment]
For anyone else who might encounter this issue, I had the same problem when I was configuring a display-only X server, with no keyboard or mouse devices listed in the Xorg config file, and using Synergy to control it from another machine.  It seems that if there are no input devices then although Awesome works initially, it very quickly loses the ability to receive keyboard or mouse events.  (Running xev shows that the events are still delivered however, Awesome just can't see them.)  Typically reloading the config and opening and closing a terminal is enough to trigger it.

I ended up assigning the default keyboard and mouse to the X server even though it wasn't being used and the problem seems to have gone away.
